File: mycuration/src/test/java/com/phicdy/mycuration/presenter/CurationListPresenterTest.java
package com.phicdy.mycuration.presenter;

import com.phicdy.mycuration.db.DatabaseAdapter;
import com.phicdy.mycuration.rss.Curation;
import com.phicdy.mycuration.view.CurationListView;

import org.junit.Test;
import org.mockito.Mockito;

import java.util.ArrayList;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.when;

public class CurationListPresenterTest {

    @Test
    public void testOnCreate() {
        // For coverage
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        presenter.setView(new MockView());
        presenter.create();
    }

    @Test
    public void listIsSetAfterOnResume() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        ArrayList<Curation> curations = new ArrayList<>();
        String testName = "testCuration";
        curations.add(new Curation(1, testName));
        when(adapter.getAllCurations()).thenReturn(curations);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        assertThat(view.curations.get(0).getName(), is(testName));
    }

    @Test
    public void testOnPause() {
        // For coverage
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        presenter.setView(new MockView());
        presenter.create();
        presenter.resume();
        presenter.pause();
    }

    @Test
    public void EditActivityStartsWhenEditIsClicked() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.onCurationEditClicked(1);
        assertThat(view.startedEditCurationId, is(1));
    }

    @Test
    public void InvalidCurationDoesNotAffectWhenEditIsClicked() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.onCurationEditClicked(-1);
        assertThat(view.startedEditCurationId, is(MockView.DEFAULT_EDIT_ID));
    }

    @Test
    public void EmptyViewIsSetIfEmptyRss() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        when(adapter.getNumOfFeeds()).thenReturn(0);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.activityCreated();
        assertThat(view.startedEditCurationId, is(MockView.DEFAULT_EDIT_ID));
    }

    @Test
    public void CurationIsDeletedWhenDeleteIsClicked() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        ArrayList<Curation> curations = new ArrayList<>();
        Curation curation = new Curation(1, "test");
        curations.add(curation);
        when(adapter.getAllCurations()).thenReturn(curations);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.onCurationDeleteClicked(curation);
        assertThat(view.curations.size(), is(0));
    }

    @Test
    public void NoRssViewIsSetWhenRssIsEmpty() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        when(adapter.getNumOfFeeds()).thenReturn(0);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.activityCreated();
        assertTrue(view.isNoRssViewSet);
    }

    @Test
    public void EmptyViewIsSetWhenRssIsEmpty() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        when(adapter.getNumOfFeeds()).thenReturn(0);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        presenter.activityCreated();
        assertTrue(view.isEmptyViewToList);
    }

    @Test
    public void Under0PositionReturnsMinusOneWhenGetCurationId() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        assertThat(presenter.getCurationIdAt(-1), is(-1));
    }

    @Test
    public void BiggerIndexThanViewSizeReturnsMinusOneWhenGetCurationId() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        ArrayList<Curation> curations = new ArrayList<>();
        String testName = "testCuration";
        curations.add(new Curation(1, testName));
        when(adapter.getAllCurations()).thenReturn(curations);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        assertThat(presenter.getCurationIdAt(curations.size()), is(-1));
    }

    @Test
    public void CurationOfIndexWhenGetCurationId() {
        DatabaseAdapter adapter = Mockito.mock(DatabaseAdapter.class);
        ArrayList<Curation> curations = new ArrayList<>();
        int testId = 1;
        curations.add(new Curation(testId, "testName"));
        when(adapter.getAllCurations()).thenReturn(curations);
        CurationListPresenter presenter = new CurationListPresenter(adapter);
        MockView view = new MockView();
        presenter.setView(view);
        presenter.create();
        presenter.resume();
        assertThat(presenter.getCurationIdAt(0), is(testId));
    }

    private class MockView implements CurationListView {
        private ArrayList<Curation> curations;
        private static final int DEFAULT_EDIT_ID = -1000;
        private int startedEditCurationId = DEFAULT_EDIT_ID;
        private boolean isNoRssViewSet = false;
        private boolean isEmptyViewToList = false;

        @Override
        public void startEditCurationActivity(int editCurationId) {
            startedEditCurationId = editCurationId;
        }

        @Override
        public void setNoRssTextToEmptyView() {
            isNoRssViewSet = true;
        }

        @Override
        public void setEmptyViewToList() {
            isEmptyViewToList = true;
        }

        @Override
        public void registerContextMenu() {

        }

        @Override
        public void initListBy(ArrayList<Curation> curations) {
            this.curations = curations;
        }

        @Override
        public void delete(Curation curation) {
            curations.remove(curation);
        }

        @Override
        public int size() {
            return 0;
        }

        @Override
        public Curation curationAt(int position) {
            return curations.get(position);
        }
    }
}

File: mycuration/src/test/java/com/phicdy/mycuration/presentation/presenter/CurationListPresenterTest.kt
package com.phicdy.mycuration.presentation.presenter

import com.phicdy.mycuration.data.db.DatabaseAdapter
import com.phicdy.mycuration.data.repository.CurationRepository
import com.phicdy.mycuration.data.repository.RssRepository
import com.phicdy.mycuration.data.repository.UnreadCountRepository
import com.phicdy.mycuration.data.rss.Curation
import com.phicdy.mycuration.presentation.view.CurationItem
import com.phicdy.mycuration.presentation.view.CurationListView
import kotlinx.coroutines.runBlocking
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import java.util.ArrayList

class CurationListPresenterTest {

    private val adapter = mock(DatabaseAdapter::class.java)
    private val view = mock(CurationListView::class.java)
    private lateinit var presenter: CurationListPresenter
    private val item = mock(CurationItem::class.java)
    private val repository = mock(UnreadCountRepository::class.java)
    private val curationRepository = mock(CurationRepository::class.java)
    private val rssRepository = mock(RssRepository::class.java)

    @Before
    fun setUp() {
        DatabaseAdapter.inject(adapter)
        presenter = CurationListPresenter(view, rssRepository, curationRepository, repository)
    }

    @Test
    fun `when after onResume then list is set`() = runBlocking {
        val curations = ArrayList<Curation>()
        val testName = "testCuration"
        curations.add(Curation(1, testName))
        `when`(curationRepository.getAllCurations()).thenReturn(curations)
        presenter.resume()
        verify(view, times(1)).initListBy(curations)
        verify(view, never()).showEmptyView()
        verify(view, times(1)).showRecyclerView()
    }

    @Test
    fun `when edit is clicked then edit activity starts`() {
        presenter.onCurationEditClicked(1)
        verify(view, times(1)).startEditCurationActivity(1)
    }

    @Test
    fun `when edit is clicked then invalid curation does not affect`() = runBlocking {
        presenter.onCurationEditClicked(-1)
        verify(view, never()).startEditCurationActivity(-1)
    }

    @Test
    fun `when empty rss then show empty view`() = runBlocking {
        `when`(curationRepository.getAllCurations()).thenReturn(arrayListOf())
        presenter.resume()
        verify(view, never()).showRecyclerView()
        verify(view, times(1)).hideRecyclerView()
        verify(view, times(1)).showEmptyView()
    }

    @Test
    fun `when delete is clicked then curation is deleted and hide the list`() = runBlocking {
        val curation = Curation(1, "test")
        val curations = arrayListOf<Curation>().apply {
            add(curation)
        }
        `when`(curationRepository.getAllCurations()).thenReturn(curations)
        presenter.onCurationDeleteClicked(curation, curations.size)
        verify(curationRepository, times(1)).delete(curation.id)
        verify(view, times(1)).hideRecyclerView()
        verify(view, times(1)).showEmptyView()
    }

    @Test
    fun `when delete is clicked in two curations then curation is deleted and not hide the list`() = runBlocking {
        val curation = Curation(1, "test")
        val curations = arrayListOf<Curation>().apply {
            add(curation)
            add(Curation(2, "test2"))
        }
        `when`(curationRepository.getAllCurations()).thenReturn(curations)
        presenter.onCurationDeleteClicked(curation, curations.size)
        verify(curationRepository, times(1)).delete(curation.id)
        verify(view, never()).hideRecyclerView()
        verify(view, never()).showEmptyView()
    }

    @Test
    fun `when rss is empty then no rss view is set`() = runBlocking {
        `when`(rssRepository.getNumOfRss()).thenReturn(0)
        presenter.activityCreated()
        verify(view, times(1)).setNoRssTextToEmptyView()
    }

    @Test
    fun `when curation is null then not set name and count`() = runBlocking {
        presenter.getView(null, item)
        verify(item, never()).setName("")
        verify(item, never()).setCount("")
    }

    @Test
    fun `when curation is not null then not set name and count`() = runBlocking {
        `when`(repository.getCurationCount(1)).thenReturn(10)
        presenter.getView(Curation(1, "name"), item)
        verify(item, times(1)).setName("name")
        verify(item, times(1)).setCount("10")
    }
}
